src/Compiler/TypedTree/TypedTreeBasics.fs

```diff
@@ -255,17 +255,26 @@ let combineNullness (nullnessOrig: Nullness) (nullnessNew: Nullness) =
 let tryAddNullnessToTy nullnessNew (ty:TType) = 
     match ty with
     | TType_var (tp, nullnessOrig) -> 
-        // TODO NULLNESS: make this avoid allocation if no change
-        Some (TType_var (tp, combineNullness nullnessOrig nullnessNew))
+        let nullnessAfter = combineNullness nullnessOrig nullnessNew
+        if nullnessAfter = nullnessOrig then
+            Some ty
+        else 
+            Some (TType_var (tp, nullnessAfter))
     | TType_app (tcr, tinst, nullnessOrig) -> 
-        // TODO NULLNESS: make this avoid allocation if no change
-        Some (TType_app (tcr, tinst, combineNullness nullnessOrig nullnessNew))
+        let nullnessAfter = combineNullness nullnessOrig nullnessNew
+        if nullnessAfter = nullnessOrig then
+            Some ty
+        else 
+            Some (TType_app (tcr, tinst, nullnessAfter))
     | TType_ucase _ -> None // TODO NULLNESS
     | TType_tuple _ -> None // TODO NULLNESS
     | TType_anon _ -> None // TODO NULLNESS
     | TType_fun (d, r, nullnessOrig) ->
-        // TODO NULLNESS: make this avoid allocation if no change
-        Some (TType_fun (d, r, combineNullness nullnessOrig nullnessNew))
+        let nullnessAfter = combineNullness nullnessOrig nullnessNew
+        if nullnessAfter = nullnessOrig then
```

Comment:
I'm not totally sure if structural equality can be used on this type. Might be best to make the equality operation explciit `nullnessEq`

Reply:
Good catch.
Will also mark Nullness and NullnessVar with `[<NoComparison;NoEquality>]`